text/0000-implied-derive.md

```diff
@@ -0,0 +1,425 @@
+`#[derive(T)]` will now imply `#[derive(C)]` for all `C` in the transitive
+super-trait closure of `T`.
+
+### Union, not concatenation
+
+When the compiler is dealing with the union of all explicitly derived and
+implicitly derived traits, the union of those traits will be derived instead of
+the concatenation. This is particularly important for two reasons:
+
+1. To minimize the breakage in transitioning to edition 2018.
+   If you in the future do `#[derive(Copy, Clone)]`,
+   then the deriving system will see this as `#[derive(Copy, Clone)]`
+   and not `#[derive(Copy, Clone, Clone)]`.
+
+2. `Eq` and `PartialOrd` both imply `PartialEq`, and so if the concatenation is
+   used rather than the union, then the compiler will derive `PartialEq` twice
+   which will result in overlapping impls.
+
+### `#[structural_match]`
+
+[RFC 1445]: https://github.com/rust-lang/rfcs/pull/1445
+
+This RFC supercedes [RFC 1445] in that `#[derive(Eq)]`, where transitively
+applied to the transitive closure of all field types of a type, implies
+`#[structural_match]`. In other words, the requirement of transitively having
+`#[derive(PartialEq, Eq)]` is reduced to `#[derive(Eq)]`.
+
+### Interaction with `#[derive_no_bound]` et al.
+
+[RFC 2353]: https://github.com/rust-lang/rfcs/pull/2353
+
+To ensure consistency, `#[derive_no_bound(Trait)]` as described in [RFC 2353]
+should also imply `#[derive_no_bound(Super)]` where `Super` is the
+transitive super-trait closure of `Trait`. Conversely,
+`#[derive_field_bound(Trait)]` will should imply `#[derive_field_bound(Super)]`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+There are two main drawbacks of this proposal.
+
+## Breakage
+
+This RFC will break some code. Specifically, if a user has already manually
+given an `impl` for a trait in the transitive closure of super-traits for a
+particular trait, then an error will be raised. An example of this situation is:
+
+```rust
+#[derive(Copy)]
+struct Foo;
+
+impl Clone for Foo {
+    fn clone(&self) -> Self { Self {} }
+}
+```
+
+which, in the future, would be equivalent to:
+
+```rust
+#[derive(Copy, Clone)]
+struct Foo;
+
+impl Clone for Foo {
+    fn clone(&self) -> Self { Self {} }
+}
+```
+
+and thus resulting in the error:
+
+```rust
+error[E0119]: conflicting implementations of trait `std::clone::Clone` for type `Foo`:
+ --> src/main.rs:2:20
+  |
+2 |     #[derive(Copy, Clone)]
+  |                    ^^^^^ conflicting implementation for `Foo`
+...
+5 |     impl Clone for Foo {
+  |     ------------------ first implementation here
+```
+
+### Mitigating factors
+
+1. We can do this breakage as part of edition 2018.
+
+2. We can give good error messages and help users to migrate with `rustfix`.
+
+3. It is expected that the breakage will be relatively small because situations
+   where `Copy` is derived but `Clone` is implemented is rare.
+   Furthermore, it `Ord` it could be downright risky to derive `Ord` but
```

Comment:
Why? If I have a struct with some super internal private field which should not be taken into account either for ordering or equality, why is this dangerous?

Reply:
> Implementations of PartialEq, PartialOrd, and Ord must agree with each other. It's easy to accidentally make them disagree by deriving some of the traits and manually implementing others.

https://doc.rust-lang.org/nightly/std/cmp/trait.PartialOrd.html#how-can-i-implement-partialord

It occurs to me that in the case the super internal private field which should not be accounted either for ordering or equality, you will manually impl all the above traits and you should, because the behavior becomes much more clear then.

Also, cc @nikomatsakis on this.

Reply:
Deriving PartialOrd or Ord (but not both) is risky because they must match or things will go awry; if you (e.g.) implement PartialOrd manually to ignore a field, but you derive Ord, then Ord will not ignore the field. This is why clippy, for example, has a lint against this -- or I think it does? At least there is an issue requesting one https://github.com/rust-lang-nursery/rust-clippy/issues/1621 =)